kotlin/src/main/kotlin/com/linecorp/armeria/internal/common/kotlin/FlowResponseConverterFunction.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2021 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.internal.common.kotlin
+
+import com.linecorp.armeria.common.HttpHeaders
+import com.linecorp.armeria.common.HttpResponse
+import com.linecorp.armeria.common.ResponseHeaders
+import com.linecorp.armeria.server.ServiceRequestContext
+import com.linecorp.armeria.server.annotation.ResponseConverterFunction
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.filterNotNull
+
+class FlowResponseConverterFunction(
+    private val responseConverter: ResponseConverterFunction
+) : ResponseConverterFunction {
+    override fun convertResponse(
+        ctx: ServiceRequestContext,
+        headers: ResponseHeaders,
+        result: Any?,
+        trailers: HttpHeaders
+    ): HttpResponse {
+        if (result is Flow<*>) {
+            // Reactive Streams doesn't allow emitting null value.
+            val publisher = asPublisher(result.filterNotNull(), ctx.eventLoop(), ctx)
+            return responseConverter.convertResponse(ctx, headers, publisher, trailers)
+        }
```

Comment:
I thought we can create a `StreamMessage` simply collect a `Flow` in a suspend function.
```kt
return when (result) {
    is Flow<*> -> {
        val future: CompletableFuture<StreamMessage<Any>> = GlobalScope.future(coroutineContext) {
            val stream = DefaultStreamMessage<Any>()
            launch {
                try {
                    result.collect { stream.write(it) }
                    stream.close()
                } catch (e: Exception) {
                    stream.close(e)
                }
            }
            stream
        }

        return HttpResponse.from(future.handle { stream, cause ->
            responseConverter.convertResponse(ctx, headers, stream, trailers)
        })
    }
    else -> ResponseConverterFunction.fallthrough()
}
```
I'm not a Kotlin expert. 😅 So it could not be the best solution.

Reply:
That was the reason why I asked `kotlinx-coroutines-reactive` could be removed.

Reply:
> the performance as a Publisher implementation itself is far less than it.

Did `less` mean coroutines-reactive's `asPublisher` showed better performance?

Reply:
> Did less mean coroutines-reactive's asPublisher showed better performance?

Yeah, if I did it correctly.
But, they showed no difference when they are integrated into `AnnotatedService`.

I will run the benchmark again and share the result.

Reply:
Sorry for the late update, but here's the result!
#### Publisher BM
```
FlowCollectingPublisherBenchmark.asPublisher              thrpt    5  5780.697 ± 828.643  ops/s
FlowCollectingPublisherBenchmark.flowCollectingPublisher  thrpt    5  1223.182 ±  23.545  ops/s
```

#### AnnotatedService BM
```
— FlowCollectingPublisher —
Benchmark                                           Mode  Cnt    Score   Error  Units
AnnotatedServiceFlowBenchmark.flowReturnType       thrpt    5  150.606 ± 5.993  ops/s
AnnotatedServiceFlowBenchmark.publisherReturnType  thrpt    5  158.658 ± 4.289  ops/s

— coroutines-reactive —
Benchmark                                           Mode  Cnt    Score    Error  Units
AnnotatedServiceFlowBenchmark.flowReturnType       thrpt    5  151.756 ±  7.422  ops/s
AnnotatedServiceFlowBenchmark.publisherReturnType  thrpt    5  149.482 ± 11.765  ops/s
```

#### BM Code:
* https://github.com/ks-yim/armeria/blob/kotlin-flow-bm/benchmarks/src/jmh/kotlin/com/linecorp/armeria/internal/common/kotlin/AnnotatedServiceFlowBenchmark.kt
* https://github.com/ks-yim/armeria/blob/kotlin-flow-bm/benchmarks/src/jmh/kotlin/com/linecorp/armeria/internal/common/kotlin/AnnotatedServiceFlowBenchmark.kt